File: components/maps/src/main/java/org/storm/maps/Map.java
package org.storm.maps;

import javafx.scene.canvas.GraphicsContext;
import lombok.Getter;
import org.storm.core.render.Renderable;
import org.storm.core.ui.Resolution;
import org.storm.core.update.Updatable;
import org.storm.maps.layer.Layer;
import org.storm.physics.entity.Entity;
import org.storm.physics.math.geometry.Point;

import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.Set;

public class Map implements Renderable, Updatable {

  private final Set<Layer> layers;

  @Getter
  private final Set<Entity> entities;

  private final Point renderPoint;

  private final Resolution resolution;

  public Map(Resolution resolution) {
    this.renderPoint = new Point(0, 0); // Top left corner of the screen
    this.layers = new LinkedHashSet<>();
    this.entities = new HashSet<>();
    this.resolution = resolution;
  }

  public void addLayer(Layer layer) {
    layer.setResolution(this.resolution);
    this.layers.add(layer);

    if (layer.isActive()) {
      this.entities.addAll(layer.getEntities());
    }
  }

  public void removeLayer(Layer layer) {
    this.layers.remove(layer);

    if (layer.isActive()) {
      this.entities.removeAll(layer.getEntities());
    }
  }

  public void setResolution(Resolution resolution) {
    this.layers
      .parallelStream()
      .forEach(layer -> layer.setResolution(resolution));
  }

  public void shiftRenderPoint(double dx, double dy) {
    this.renderPoint.translate(dx, dy);
  }

  @Override
  public void render(GraphicsContext graphicsContext, double x, double y) {
    this.layers
      .forEach(layer -> layer.render(graphicsContext, renderPoint.getX(), renderPoint.getY()));
  }

  @Override
  public void update(double time, double frameTime) {
    for (Layer layer : this.layers) {
      if (layer.isActive()) {
        layer.update(time, frameTime);
      }
    }
  }

}

File: components/maps/src/main/kotlin/org/storm/maps/Map.kt
package org.storm.maps

import org.storm.core.render.canvas.Canvas
import org.storm.core.render.Renderable
import org.storm.core.render.geometry.Point
import org.storm.core.update.Updatable
import org.storm.maps.layer.Layer
import org.storm.physics.entity.Entity

class Map : Renderable, Updatable {

    // Top left corner of the screen
    private val renderPoint: Point = Point(0.0, 0.0)

    private val layers: MutableSet<Layer> = LinkedHashSet()

    val entities: MutableSet<Entity> = mutableSetOf()

    fun addLayer(layer: Layer) {
        layers.add(layer)

        if (layer.active) {
            entities.addAll(layer.entities)
        }
    }

    fun removeLayer(layer: Layer) {
        layers.remove(layer)

        if (layer.active) {
            entities.removeAll(layer.entities)
        }
    }

    fun shiftRenderPoint(dx: Double, dy: Double) {
        renderPoint.translate(dx, dy)
    }

    override suspend fun render(canvas: Canvas, x: Double, y: Double) {
        layers.forEach { layer -> layer.render(canvas, renderPoint.x, renderPoint.y) }
    }

    override suspend fun update(time: Double, elapsedTime: Double) {
        layers.filter {
            it.active
        }.forEach {
            it.update(time, elapsedTime)
        }
    }
}
